File: 2020/1337.the-k-weakest-rows-in-a-matrix.js
/*
 * @lc app=leetcode id=1337 lang=javascript
 *
 * [1337] The K Weakest Rows in a Matrix
 */

// @lc code=start
/**
 * @param {number[][]} mat
 * @param {number} k
 * @return {number[]}
 */
var kWeakestRows = function(mat, k) {
  let hash = {} // indexes, indexed by strength
  mat.forEach((row,i)=>{
    let strength = row.lastIndexOf(1)
    hash[strength] = [...(hash[strength]??[]), i]
  })
  let sorted = Object.entries(hash).sort((a,b)=>a[0]-b[0])
  sorted = sorted.map(v=>v[1])
  return sorted.flat().slice(0,k)
};
// @lc code=end

File: 1337.the-k-weakest-rows-in-a-matrix.py
#
# @lc app=leetcode id=1337 lang=python3
#
# [1337] The K Weakest Rows in a Matrix
#

# @lc code=start
class Solution:
    def kWeakestRows(self, mat: List[List[int]], k: int) -> List[int]:
        mapd = {}
        for i, row in enumerate(mat):
            n = row.count(1)
            if n not in mapd:
                mapd[n] = []
            mapd[n].append(i)

        keys = sorted(mapd.keys())
        vals = [idx for key in keys for idx in mapd[key]]
        return vals[:k]

# @lc code=end

